tsdb/chunks/head_chunks.go

```diff
@@ -0,0 +1,786 @@
+}
+
+// ChunkDiskMapper is for writing the Head block chunks to the disk
+// and access chunks via mmapped file.
+type ChunkDiskMapper struct {
+	// Writer.
+	dir     *os.File
+	curFile *os.File // File being written to.
+
+	curFileMint     int64 // In milliseconds.
+	curFileMaxt     int64 // In milliseconds.
+	curFileSequence int
+	n               int64 // Bytes written in current open file.
+	maxFileMs       int64
+
+	writeBuf     [8]byte
+	wbuf         *bufio.Writer
+	crc32        hash.Hash
+	writePathMtx sync.RWMutex
+
+	// Reader.
+	// The int key in the map is the file number on the disk.
+	mmappedChunkFiles map[int]ByteSlice
+	closers           map[int]io.Closer // Closers for resources behind the byte slices.
+	pool              chunkenc.Pool
+	readPathMtx       sync.RWMutex
+	chunkBuffer       *chunkBuffer
+
+	size   int64 // The total size of bytes in the reader.
+	closed bool
+}
+
+const (
+	// MintMaxtSize is the size of the mint/maxt for head chunk file and chunks.
+	MintMaxtSize = 8
+	// SeriesRefSize is the size of series reference on disk.
+	SeriesRefSize = 8
+	// HeadChunkFileHeaderSize is the total size of the header for the head chunk file.
+	HeadChunkFileHeaderSize = SegmentHeaderSize + 2*MintMaxtSize
+	// HeaderMintOffset is the offset where the first byte of MinT for head chunk file exists.
+	HeaderMintOffset = SegmentHeaderSize
+	// HeaderMaxtOffset is the offset where the first byte of MaxT for head chunk file exists.
+	HeaderMaxtOffset = HeaderMintOffset + MintMaxtSize
+	// MaxHeadChunkFileSize is the max size of a head chunk file.
+	MaxHeadChunkFileSize = 1 * 1024 * 1024 * 1024 // 1 GiB.
+	// CRCSize is the size of crc32 sum on disk.
+	CRCSize = 4
+	// MaxHeadChunkMetaSize is the max size of an mmapped chunks minus the chunks data.
+	// Max because the uvarint size can be smaller.
+	MaxHeadChunkMetaSize = SeriesRefSize + 2*MintMaxtSize + ChunksFormatVersionSize + MaxChunkLengthFieldSize + CRCSize
+)
+
+// NewChunkDiskMapper returns a new writer against the given directory
+// using the default head chunk file duration.
+func NewChunkDiskMapper(dir string, pool chunkenc.Pool) (*ChunkDiskMapper, error) {
+	return newChunkDiskMapper(dir, DefaultHeadChunkFileMaxTimeRange, pool)
+}
+
+func newChunkDiskMapper(dir string, maxFileDuration int64, pool chunkenc.Pool) (*ChunkDiskMapper, error) {
+	if maxFileDuration <= 0 {
+		maxFileDuration = DefaultHeadChunkFileMaxTimeRange
+	}
+
+	if err := os.MkdirAll(dir, 0777); err != nil {
+		return nil, err
+	}
+	dirFile, err := fileutil.OpenDir(dir)
+	if err != nil {
+		return nil, err
+	}
+
+	m := &ChunkDiskMapper{
+		dir:         dirFile,
+		maxFileMs:   maxFileDuration,
+		pool:        pool,
+		crc32:       newCRC32(),
+		chunkBuffer: newChunkBuffer(),
+	}
+
+	return m, m.openMMapFiles()
+}
+
+func (cdm *ChunkDiskMapper) openMMapFiles() (returnErr error) {
+	bs := map[int]ByteSlice{}
+	cs := map[int]io.Closer{}
```

Comment:
Did you forget to comment or push the changes? This and few other places were marked as resolved without any comment or code changes/

Reply:
I tend to mark as resolve as and when I am coding it locally so that it is easier to keep track of unresolved comments (when we have so many comments). I will be pushing the updates shortly.